### PR TITLE
Initial Gift development

### LINKED ISSUE #5463: BLB: Gift Keyword
Two Version, one for Permanent Spells and one for instant or sorcery spell
> As you cast a permanent spell with gift, you can promise the gift to an opponent if you want. This is optional, and if you have more than one opponent—ooh, scary!—you choose one and only one of them at most. They don't receive the gift yet. You're just promising it. When the permanent enters, the chosen opponent receives the gift. In Kitnap's case, they draw a card.
> Each card with gift will have additional or alternative effects if the gift was or wasn't promised. In Kitnap's case, promising the gift will change how its other enters-the-battlefield triggered ability functions. If you agree to let an opponent draw a card, you can avoid putting three stun counters on the creature you just … um, adopted.
> But not every gift is a card, and not every card with gift is a permanent. Let's look at an instant and a Fish.
> Promising the gift for an instant or sorcery spell works just like it does for a permanent spell. You do this as you cast the spell. It's an additional cost. The opponent receiving the gift is a little different. As the instant or sorcery spell starts resolving, first the opponent receives their gift. In Parting Gust's case, they create a tapped 1/1 blue Fish creature token. Yay! A free snack and/or pet and/or good luck charm. Then the spell's other effects happen.

The "Gift a tapped fish" Token uses a Named one, so LockTokenScript and no ChangeCardText

Other Gift Effects:

- Gift a Treasure
- Gift a card
- Gift an extra Turn

Also, "Whenever you gift" Trigger

### PR DESCRIPTION
**Pull request by tehdiplomat**
Closes #5463

### CODE REVIEW

File: forge-game/src/main/java/forge/game/card/CardFactoryUtil.java [hunk 1]
```diff
@@ -1255,6 +1225,22 @@ public static void addTriggerAbility(final KeywordInterface inst, final Card car
 
             saRebel.setIntrinsic(intrinsic);
             inst.addTrigger(etbTrigger);
+        } else if (keyword.equals("Gift")) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/card/CardFactoryUtil.java:1228]: The ETB trigger only makes sense for Permanents, so i think we could skip them for others?

**Review by Agetian — APPROVED** [review]

**Comment by Agetian** [comment]
Probably worth updating GameState to support |Promised for the save state and puzzle purposes, but I can take care of it later once this PR is in.

File: forge-game/src/main/java/forge/game/ability/AbilityUtils.java [hunk 2]
```diff
@@ -1352,6 +1354,28 @@ public static void resolve(final SpellAbility sa) {
...
+        Player controller = sa.getActivatingPlayer();
+        Card source = sa.getHostCard();
+        // do blessing there before condition checks
+        if (sa.isSpell() && !source.isPermanent() && source.hasKeyword(Keyword.ASCEND)) {
+            if (controller.getZone(ZoneType.Battlefield).size() >= 10) {
+                controller.setBlessing(true);
+            }
+        }
+
+        if (sa.isSpell() && !(sa instanceof SpellPermanent) && sa.getHostCard().hasKeyword(Keyword.GIFT) && sa.isOptionalCostPaid(OptionalCost.PromiseGift)) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/AbilityUtils.java:1367]: I don't know if `!source.isPermanent()` wouldn't be better like above with Ascend

**Comment by Hanmac** [comment]
I probably would have written it like `K:Gift:GiftEffect`, but that might be changed later

File: forge-game/src/main/java/forge/game/card/Card.java [hunk 3]
```diff
@@ -1888,6 +1889,18 @@ public final void setChosenPlayer(final Player p) {
         view.updateChosenPlayer(this);
     }
 
+    public final boolean hasPromisedGift() {
+        return promisedGift != null;
+    }
+    public final Player getPromisedGift() {
+        return promisedGift;
+    }
+    public final void setPromisedGift(final Player p) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/Card.java:1898]: this needs to be called in `copySpellHost`

File: forge-game/src/main/java/forge/game/card/CardFactoryUtil.java [hunk 4]
```diff
@@ -1255,6 +1225,22 @@ public static void addTriggerAbility(final KeywordInterface inst, final Card car
...
+            // Gift is a special keyword that is used to create a trigger for permanents when they enter the battlefield
+            // On casting a Gift needs to be promised to an opponent
+            final SpellAbility saGift = AbilityFactory.getAbility(card.getSVar("GiftEffect"), card);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/card/CardFactoryUtil.java:1231]: What happens when you do this for non-Permanents?

`card.getFirstSpellAbility().setAdditionalAbility("GiftAbility", saGift)` ?

File: forge-game/src/main/java/forge/game/ability/AbilityUtils.java (hunk 2 above)
Comment (tehdiplomat) [reply forge-game/src/main/java/forge/game/ability/AbilityUtils.java:1367]: Is source.isPermanent right? Does that calculate properly for things like Adventure?

File: forge-game/src/main/java/forge/game/card/CardFactoryUtil.java (hunk 1 above)
Comment (tehdiplomat) [reply forge-game/src/main/java/forge/game/card/CardFactoryUtil.java:1228]: Yea i know. I wasn't done with this part yet.

File: forge-game/src/main/java/forge/game/ability/AbilityUtils.java (hunk 2 above)
Comment (Hanmac) [reply forge-game/src/main/java/forge/game/ability/AbilityUtils.java:1367]: At this place, shortly before the Spell is going to resolve, 
the Spell Host has changed into the Adventure Spell Half (so it is not a Permanent, and it shouldn't have had Gift anyway)

File: forge-game/src/main/java/forge/game/card/Card.java (hunk 3 above)
Comment (tehdiplomat) [reply forge-game/src/main/java/forge/game/card/Card.java:1898]: Lemme know if thats what you meant

**Review by Hanmac — APPROVED** [review]

**Review by Hanmac — APPROVED** [review]

**Review by tool4ever — APPROVED** [review]

**Pull request merged**